src/Processors/Formats/Impl/ValuesBlockInputFormat.cpp

```diff
@@ -181,6 +184,57 @@ bool ValuesBlockInputFormat::tryReadValue(IColumn & column, size_t column_idx)
     }
 }
 
+namespace
+{
+    void tryToReplaceNullFieldsInTupleOrArrayWithDefaultValues(Field & value, const IDataType & type)
+    {
+        checkStackSize();
+
+        const DataTypeTuple * type_tuple = typeid_cast<const DataTypeTuple *>(&type);
+        const DataTypeArray * type_array = typeid_cast<const DataTypeArray *>(&type);
+
+        if (type_tuple && value.getType() == Field::Types::Tuple)
+        {
+            Tuple & tuple_value = value.get<Tuple>();
+
+            size_t src_tuple_size = tuple_value.size();
+            size_t dst_tuple_size = type_tuple->getElements().size();
+
+            if (src_tuple_size != dst_tuple_size)
+                throw Exception(fmt::format("Bad size of tuple. Expected size: {}, actual size: {}.",
+                    std::to_string(src_tuple_size), std::to_string(dst_tuple_size)), ErrorCodes::TYPE_MISMATCH);
+
+            for (size_t i = 0; i < src_tuple_size; ++i)
+            {
+                const auto & element_type = *(type_tuple->getElements()[i]);
+
+                if (tuple_value[i].isNull() && !element_type.isNullable())
+                    tuple_value[i] = element_type.getDefault();
+
+                tryToReplaceNullFieldsInTupleOrArrayWithDefaultValues(tuple_value[i], element_type);
+            }
+        }
+        else if (type_array && value.getType() == Field::Types::Array)
```

Comment:
The code is not generic, for example, it does not work for recently added Map data type.
Not sure if it's easy to fix.